conans/client/tools/win.py

```diff
@@ -63,22 +64,71 @@ def build_sln_command(settings, sln_path, targets=None, upgrade_project=True, bu
     return command
 
 
-def vs_installation_path(version):
+def vs_installation_path(version, preference=None):
+
+    if not preference:
+        preference = [var.lstrip().rstrip() for var in get_env("CONAN_VS_INSTALLATION_PREFERENCE",
+                                                               list())]
+    
+    if not preference:
+        preference = ["Enterprise", "Professional", "Community", "BuildTools"]
+
     if not hasattr(vs_installation_path, "_cached"):
         vs_installation_path._cached = dict()
 
-    if version not in vs_installation_path._cached:
-        vs_path = None
-        legacy_products = vswhere(legacy=True)
+    if version in vs_installation_path._cached:
+        if vs_installation_path._cached[version] is not None:
+            if preference[0] in vs_installation_path._cached[version]:
+                return vs_installation_path._cached[version]
+            else:
+                vs_installation_path._cached[version] = None
+
+    # Try with vswhere()
+    products = None
+    try:
+        legacy_products =vswhere(legacy=True)
         all_products = vswhere(products=["*"])
-        products = legacy_products + all_products
+        products = list(legacy_products + all_products)
+    except ConanException:
+        pass
+
+    vs_paths = list()
 
+    if products:
+        # remove repeated products
+        seenProducts = list()
```

Comment:
Oneline (note ``seen_products``):
```python
# if order of vswhere output is important, if not, use set() instead of OrderedDict
seen_products = list(OrderedDict.fromkeys(products))
```
Maybe doesn't work if products is not hashable, are they string or dict?